src/main/integrations/renderer-ipc.ts

```diff
@@ -0,0 +1,114 @@
+
+  /** @inheritDoc */
+  public name: string = RendererIPC.id;
+
+  /** @inheritDoc */
+  public setupOnce(): void {
+    this._setupIPC();
+  }
+
+  /** Sets up IPC from renderers */
+  private _setupIPC(): void {
+    const options = getCurrentHub().getClient<NodeClient>()?.getOptions() as ElectronMainOptions | undefined;
+
+    ipcMain.on(IPC.EVENT, (event: IpcMainEvent, jsonEvent: string) =>
+      this._handleEvent(jsonEvent, event.sender, options),
+    );
+
+    ipcMain.on(IPC.SCOPE, (_: any, jsonScope: string) => this._handleScope(jsonScope, options));
+    ipcMain.on(IPC.CONTEXT, (event) => this._handleContext(event.sender, options));
+  }
+
+  /**
+   * Handle events from the renderer processes
+   */
+  private _handleEvent(jsonEvent: string, contents: WebContents, options?: ElectronMainOptions): void {
+    let event: Event;
+    try {
+      event = JSON.parse(jsonEvent) as Event;
+    } catch {
+      logger.warn('sentry-electron received an invalid IPC_EVENT message');
+      return;
+    }
+
+    event.release = options?.release;
+
+    if (event.exception) {
+      event.contexts = {
+        electron: {
+          crashed_process: options?.getRendererName?.(contents) || `WebContents[${contents.id}]`,
+          crashed_url: normalizeUrl(contents.getURL(), app.getAppPath()),
+        },
+        ...event.contexts,
+      };
+    }
+
+    captureEvent(event);
+  }
+
+  /**
+   * Handle scope updates from renderer processes
+   */
+  private _handleScope(jsonScope: string, options?: ElectronMainOptions): void {
+    let rendererScope: Scope;
+    try {
+      rendererScope = JSON.parse(jsonScope) as Scope;
+    } catch {
+      logger.warn('sentry-electron received an invalid IPC_SCOPE message');
+      return;
+    }
+
+    const sentScope = Scope.clone(rendererScope) as any;
+    /* eslint-disable @typescript-eslint/no-unsafe-member-access */
+    configureScope((scope) => {
+      if (hasKeys(sentScope._user)) {
+        scope.setUser(sentScope._user);
+      }
+
+      if (hasKeys(sentScope._tags)) {
+        scope.setTags(sentScope._tags);
+      }
+
+      if (hasKeys(sentScope._extra)) {
+        scope.setExtras(sentScope._extra);
+      }
+
+      scope.addBreadcrumb(sentScope._breadcrumbs.pop(), options?.maxBreadcrumbs || 100);
+    });
+    /* eslint-enable @typescript-eslint/no-unsafe-member-access */
+  }
+
+  /**
+   * Gets the Electron context and passes it to the renderer
+   *
+   * This allows the browser SDK to be used in isolation and still include Electron context and normalise paths
+   */
+  private async _handleContext(contents: WebContents, options?: ElectronMainOptions): Promise<void> {
+    const eventDefaults = await getEventDefaults(options?.release);
+
+    const crashed_process = options?.getRendererName?.(contents) || `WebContents[${contents.id}]`;
+    eventDefaults.contexts = { ...eventDefaults.contexts, electron: { ...eventDefaults.contexts, crashed_process } };
```

Comment:
> In the browser test, there appears to be repeated data in both context.app and context.electron.app

`electron: { ...eventDefaults.contexts, crashed_process }`

Is this why this occurs?

Reply:
> Is this why this occurs?

Yes it is, thanks! 🤦‍♂️ 

> We can add a counter to check for this for now?

The tests [already wait for a specific number of events](https://github.com/getsentry/sentry-electron/pull/361/files#diff-e5e51645622411520158296453f2a672555905fc58242d089a42ccf582ed4fe5R121). We probably need to wait a couple more seconds to ensure there aren't any **extra** events after the expected count. This might be how they're sneaking into subsequent tests. Most of the tests use the same app with the a single name and this is used by Electron to locate the minidump directory.

I've got an [unfinished branch](https://github.com/timfish/sentry-electron/tree/recipes) which moves the tests into individual markdown files (which I've called recipes 🤷‍♂️) and the bottom of each recipe contains any number of [normalised](https://github.com/timfish/sentry-electron/blob/recipes/test/e2e/normalize.ts) events which must match what the server receives. 

For example, [this recipe](https://github.com/timfish/sentry-electron/blob/recipes/test/e2e/recipes/electron-forge.md) is a simple [Electron Forge](https://electronforge.io) app and at the bottom is the expected event data. 

[Here](https://github.com/timfish/sentry-electron/blob/recipes/test/e2e/recipes/webpack-context-isolation.md) is a basic Electron application built with webpack 5 with its expected event.

This has quite a few positives:
- Each test recipe is self contained and also tests bundling success/failure
- Each test app can have a unique app name (so no minidump leakage between tests)
- Resulting event data is more thoroughly compared to the expected events
- Each recipe serves as a complete example of using the SDK
- Each test uses the [packed .tgz output](https://github.com/timfish/sentry-electron/blob/fd9261ec9050b21b067ae17928d1ec1602ca0447/test/e2e/recipe-runner.ts#L88-L93) rather than a [relative path](https://github.com/getsentry/sentry-electron/blob/117c55f36cef1f25b045e37e06425b067767861e/test/e2e/test-app/fixtures/sentry-basic.js#L1) so we know the publish output is good

Reply:
> The tests already wait for a specific number of events

Gotcha, yeah then the timeout is probably the right idea here.

Damn those recipes are super cool. You think we can extract this into it's own package and use it with https://github.com/getsentry/sentry-react-native or https://github.com/getsentry/sentry-javascript/tree/master/packages/nextjs? (I will gladly help make this happen 😄)?